File: Context/Sources/CodeFire/Services/ImageGenerationService.swift
import Foundation
import GRDB

actor ImageGenerationService {

    struct GenerationResult {
        let imageData: Data
        let responseText: String?
        let error: String?
    }

    private let model = "google/gemini-3.1-flash-image-preview"
    private let endpoint = "https://openrouter.ai/api/v1/chat/completions"

    // Per-session conversation history for multi-turn editing
    private var conversationHistory: [[String: Any]] = []

    /// Generate an image from a text prompt.
    func generateImage(
        prompt: String,
        aspectRatio: String = "1:1",
        imageSize: String = "1K"
    ) async -> GenerationResult {
        let message: [String: Any] = [
            "role": "user",
            "content": [["type": "text", "text": prompt]]
        ]
        conversationHistory.append(message)
        return await callAPI(aspectRatio: aspectRatio, imageSize: imageSize)
    }

    /// Edit an existing image with a text prompt (image-to-image).
    func editImage(
        imageData: Data,
        prompt: String,
        aspectRatio: String = "1:1",
        imageSize: String = "1K"
    ) async -> GenerationResult {
        let base64 = imageData.base64EncodedString()
        let dataURL = "data:image/png;base64,\(base64)"
        let message: [String: Any] = [
            "role": "user",
            "content": [
                ["type": "text", "text": prompt],
                ["type": "image_url", "image_url": ["url": dataURL]]
            ]
        ]
        conversationHistory.append(message)
        return await callAPI(aspectRatio: aspectRatio, imageSize: imageSize)
    }

    /// Clear conversation history to start a fresh session.
    func resetConversation() {
        conversationHistory = []
    }

    // MARK: - API Call

    private func callAPI(
        aspectRatio: String,
        imageSize: String
    ) async -> GenerationResult {
        let apiKey = await MainActor.run { ClaudeService.openRouterAPIKey }
        guard let apiKey, !apiKey.isEmpty else {
            return GenerationResult(imageData: Data(), responseText: nil, error: "OpenRouter API key not configured")
        }

        let body: [String: Any] = [
            "model": model,
            "modalities": ["image", "text"],
            "messages": conversationHistory,
            "image_config": [
                "aspect_ratio": aspectRatio,
                "image_size": imageSize
            ]
        ]

        guard let bodyData = try? JSONSerialization.data(withJSONObject: body) else {
            return GenerationResult(imageData: Data(), responseText: nil, error: "Failed to encode request")
        }

        var request = URLRequest(url: URL(string: endpoint)!)
        request.httpMethod = "POST"
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("CodeFire", forHTTPHeaderField: "X-Title")
        request.httpBody = bodyData
        request.timeoutInterval = 120 // Image generation can take 10-30s

        do {
            let (data, response) = try await URLSession.shared.data(for: request)

            guard let httpResponse = response as? HTTPURLResponse else {
                return GenerationResult(imageData: Data(), responseText: nil, error: "Invalid response")
            }

            guard httpResponse.statusCode == 200 else {
                let raw = String(data: data, encoding: .utf8) ?? "Unknown"
                return GenerationResult(imageData: Data(), responseText: nil, error: "HTTP \(httpResponse.statusCode): \(String(raw.prefix(300)))")
            }

            guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                return GenerationResult(imageData: Data(), responseText: nil, error: "Invalid JSON response")
            }

            if let error = json["error"] as? [String: Any],
               let message = error["message"] as? String {
                return GenerationResult(imageData: Data(), responseText: nil, error: message)
            }

            // Parse response
            guard let choices = json["choices"] as? [[String: Any]],
                  let firstChoice = choices.first,
                  let message = firstChoice["message"] as? [String: Any] else {
                let raw = String(data: data, encoding: .utf8) ?? "?"
                print("ImageGenerationService: unexpected top-level: \(String(raw.prefix(500)))")
                return GenerationResult(imageData: Data(), responseText: nil, error: "Unexpected response format")
            }

            // Check for refusal first
            if let refusal = message["refusal"] as? String, !refusal.isEmpty {
                return GenerationResult(imageData: Data(), responseText: nil, error: "Model refused: \(refusal)")
            }

            // Extract text content (may be empty string or array)
            var responseText: String?
            if let str = message["content"] as? String, !str.isEmpty {
                responseText = str
            } else if let arr = message["content"] as? [[String: Any]] {
                for part in arr {
                    if part["type"] as? String == "text", let text = part["text"] as? String, !text.isEmpty {
                        responseText = text
                    }
                }
            }

            // Extract image — OpenRouter/Gemini returns images in message["images"] array
            var imageData: Data?

            // Primary: message.images[] (OpenRouter's actual format)
            if let images = message["images"] as? [[String: Any]] {
                for img in images {
                    if let imageUrl = img["image_url"] as? [String: Any],
                       let urlString = imageUrl["url"] as? String,
                       let commaIndex = urlString.firstIndex(of: ",") {
                        let base64String = String(urlString[urlString.index(after: commaIndex)...])
                        imageData = Data(base64Encoded: base64String)
                        if imageData != nil { break }
                    }
                }
            }

            // Fallback: content array with image_url parts
            if imageData == nil, let arr = message["content"] as? [[String: Any]] {
                for part in arr {
                    if part["type"] as? String == "image_url",
                       let imageUrl = part["image_url"] as? [String: Any],
                       let urlString = imageUrl["url"] as? String,
                       let commaIndex = urlString.firstIndex(of: ",") {
                        let base64String = String(urlString[urlString.index(after: commaIndex)...])
                        imageData = Data(base64Encoded: base64String)
                        if imageData != nil { break }
                    }
                }
            }

            // Append assistant message to conversation history for multi-turn
            // Reconstruct content with images for the API
            var assistantContent: [[String: Any]] = []
            if let text = responseText {
                assistantContent.append(["type": "text", "text": text])
            }
            if let images = message["images"] as? [[String: Any]] {
                assistantContent.append(contentsOf: images)
            }
            if !assistantContent.isEmpty {
                conversationHistory.append([
                    "role": "assistant",
                    "content": assistantContent
                ])
            }

            guard let finalImageData = imageData, !finalImageData.isEmpty else {
                // Check finish_reason for clues
                let finishReason = firstChoice["native_finish_reason"] as? String
                    ?? firstChoice["finish_reason"] as? String ?? "unknown"
                print("ImageGenerationService: no image. finish_reason=\(finishReason), content=\(message["content"] ?? "nil")")

                if finishReason.lowercased().contains("safety") || finishReason.lowercased().contains("block") {
                    return GenerationResult(imageData: Data(), responseText: responseText, error: "Image blocked by safety filter. Try a different prompt.")
                }
                return GenerationResult(imageData: Data(), responseText: responseText, error: "No image generated. The model may have declined this prompt. Try rephrasing.")
            }

            return GenerationResult(imageData: finalImageData, responseText: responseText, error: nil)
        } catch {
            return GenerationResult(imageData: Data(), responseText: nil, error: error.localizedDescription)
        }
    }

    // MARK: - File Management

    /// Save image data to disk and create a DB record. Returns the saved GeneratedImage.
    /// This is a static function that runs on MainActor since it accesses DatabaseService.shared.
    @MainActor
    static func saveGeneration(
        imageData: Data,
        prompt: String,
        responseText: String?,
        projectId: String,
        projectPath: String,
        aspectRatio: String,
        imageSize: String,
        parentImageId: Int64? = nil
    ) -> GeneratedImage? {
        // Build save directory
        let saveDir: URL
        if projectPath.isEmpty {
            let appSupport = FileManager.default.urls(
                for: .applicationSupportDirectory,
                in: .userDomainMask
            ).first!.appendingPathComponent("CodeFire/generated-images", isDirectory: true)
            saveDir = appSupport
        } else {
            saveDir = URL(fileURLWithPath: projectPath)
                .appendingPathComponent("assets/generated", isDirectory: true)
        }

        do {
            try FileManager.default.createDirectory(at: saveDir, withIntermediateDirectories: true)
        } catch {
            print("ImageGenerationService: failed to create directory: \(error)")
            return nil
        }

        // Generate unique filename
        let timestamp = Int(Date().timeIntervalSince1970)
        let filename = "gen_\(timestamp)_\(UUID().uuidString.prefix(8)).png"
        let filePath = saveDir.appendingPathComponent(filename)

        do {
            try imageData.write(to: filePath)
        } catch {
            print("ImageGenerationService: failed to write image: \(error)")
            return nil
        }

        // Insert DB record
        var image = GeneratedImage(
            projectId: projectId,
            prompt: prompt,
            responseText: responseText,
            filePath: filePath.path,
            aspectRatio: aspectRatio,
            imageSize: imageSize,
            parentImageId: parentImageId,
            createdAt: Date()
        )

        do {
            try DatabaseService.shared.dbQueue.write { db in
                try image.insert(db)
            }
            return image
        } catch {
            print("ImageGenerationService: failed to insert DB record: \(error)")
            return nil
        }
    }
}

File: electron/src/main/services/ImageGenerationService.ts
import { net } from 'electron'
import fs from 'fs'
import path from 'path'
import { app } from 'electron'
import { getModelById } from '@shared/media-models'

const ENDPOINT = 'https://openrouter.ai/api/v1/chat/completions'

interface GenerationResult {
  imagePath: string | null
  responseText: string | null
  error: string | null
}

export interface GenerateOptions {
  prompt: string
  apiKey: string
  model: string
  aspectRatio?: string
  imageSize?: string
  seed?: number
  referenceImages?: string[] // file paths for img2img / style transfer / character ref
}

export class ImageGenerationService {
  private conversationHistory: Record<string, unknown>[] = []

  async generate(opts: GenerateOptions): Promise<GenerationResult> {
    if (!opts.apiKey) {
      return { imagePath: null, responseText: null, error: 'OpenRouter API key not configured' }
    }

    const contentParts: Record<string, unknown>[] = []

    // Add reference images (for img2img, style transfer, character ref)
    if (opts.referenceImages?.length) {
      for (const refPath of opts.referenceImages) {
        try {
          const imgData = fs.readFileSync(refPath)
          const ext = path.extname(refPath).toLowerCase().replace('.', '')
          const mime = ext === 'jpg' ? 'image/jpeg' : `image/${ext || 'png'}`
          const dataUrl = `data:${mime};base64,${imgData.toString('base64')}`
          contentParts.push({ type: 'image_url', image_url: { url: dataUrl } })
        } catch (err) {
          return { imagePath: null, responseText: null, error: `Failed to read reference image: ${err instanceof Error ? err.message : String(err)}` }
        }
      }
    }

    contentParts.push({ type: 'text', text: opts.prompt })

    const message = { role: 'user', content: contentParts }
    this.conversationHistory.push(message)

    return this.callAPI(opts)
  }

  async editImage(
    originalImagePath: string,
    editPrompt: string,
    opts: Omit<GenerateOptions, 'prompt'>
  ): Promise<GenerationResult> {
    if (!opts.apiKey) {
      return { imagePath: null, responseText: null, error: 'OpenRouter API key not configured' }
    }

    let imageDataUrl: string
    try {
      const imgData = fs.readFileSync(originalImagePath)
      imageDataUrl = `data:image/png;base64,${imgData.toString('base64')}`
    } catch (err) {
      return { imagePath: null, responseText: null, error: `Failed to read image: ${err instanceof Error ? err.message : String(err)}` }
    }

    this.conversationHistory = [
      {
        role: 'user',
        content: [
          { type: 'image_url', image_url: { url: imageDataUrl } },
          { type: 'text', text: editPrompt },
        ],
      },
    ]

    return this.callAPI({ ...opts, prompt: editPrompt })
  }

  resetConversation() {
    this.conversationHistory = []
  }

  private async callAPI(opts: GenerateOptions): Promise<GenerationResult> {
    // Look up model to get correct output modalities
    const modelDef = getModelById(opts.model)
    const modalities = modelDef?.outputModalities ?? ['image', 'text']

    const bodyObj: Record<string, unknown> = {
      model: opts.model,
      modalities,
      messages: this.conversationHistory,
      image_config: {
        aspect_ratio: opts.aspectRatio ?? '1:1',
        image_size: opts.imageSize ?? '1K',
      },
    }

    if (opts.seed !== undefined) {
      bodyObj.seed = opts.seed
    }

    const body = JSON.stringify(bodyObj)
    console.log('[ImageGen] REQUEST model:', opts.model, 'modalities:', modalities, 'aspectRatio:', opts.aspectRatio, 'imageSize:', opts.imageSize)

    return new Promise((resolve) => {
      const request = net.request({ method: 'POST', url: ENDPOINT })
      request.setHeader('Authorization', `Bearer ${opts.apiKey}`)
      request.setHeader('Content-Type', 'application/json')
      request.setHeader('X-Title', 'CodeFire')

      let responseData = ''

      request.on('response', (response) => {
        response.on('data', (chunk) => { responseData += chunk.toString() })

        response.on('end', () => {
          console.log('[ImageGen] RESPONSE status:', response.statusCode)
          if (response.statusCode !== 200) {
            console.log('[ImageGen] RESPONSE error body:', responseData.substring(0, 1000))
            resolve({ imagePath: null, responseText: null, error: `HTTP ${response.statusCode}: ${responseData.substring(0, 300)}` })
            return
          }

          try {
            const json = JSON.parse(responseData)

            if (json.error?.message) {
              resolve({ imagePath: null, responseText: null, error: json.error.message })
              return
            }

            const message = json.choices?.[0]?.message
            if (!message) {
              resolve({ imagePath: null, responseText: null, error: 'Unexpected response format' })
              return
            }

            if (message.refusal) {
              resolve({ imagePath: null, responseText: null, error: `Model refused: ${message.refusal}` })
              return
            }

            // Extract text
            let responseText: string | null = null
            if (typeof message.content === 'string' && message.content) {
              responseText = message.content
            } else if (Array.isArray(message.content)) {
              for (const part of message.content) {
                if (part.type === 'text' && part.text) {
                  responseText = part.text
                }
              }
            }

            // Extract image data
            let imageBase64: string | null = null

            if (Array.isArray(message.images)) {
              for (const img of message.images) {
                const url = img.image_url?.url
                if (url) {
                  const commaIdx = url.indexOf(',')
                  if (commaIdx !== -1) { imageBase64 = url.substring(commaIdx + 1); break }
                }
              }
            }

            if (!imageBase64 && Array.isArray(message.content)) {
              for (const part of message.content) {
                if (part.type === 'image_url') {
                  const url = part.image_url?.url
                  if (url) {
                    const commaIdx = url.indexOf(',')
                    if (commaIdx !== -1) { imageBase64 = url.substring(commaIdx + 1); break }
                  }
                }
              }
            }

            if (!imageBase64) {
              resolve({ imagePath: null, responseText, error: responseText ? null : 'No image returned' })
              return
            }

            // Save image to disk
            const imagesDir = path.join(app.getPath('userData'), 'generated-images')
            if (!fs.existsSync(imagesDir)) {
              fs.mkdirSync(imagesDir, { recursive: true })
            }
            const fileName = `${Date.now()}-${Math.random().toString(36).substring(2, 8)}.png`
            const filePath = path.join(imagesDir, fileName)
            fs.writeFileSync(filePath, Buffer.from(imageBase64, 'base64'))

            // Append assistant message to conversation history
            const assistantContent: Record<string, unknown>[] = []
            if (responseText) { assistantContent.push({ type: 'text', text: responseText }) }
            if (message.images) { assistantContent.push(...message.images) }
            if (assistantContent.length > 0) {
              this.conversationHistory.push({ role: 'assistant', content: assistantContent })
            }

            resolve({ imagePath: filePath, responseText, error: null })
          } catch (err) {
            resolve({ imagePath: null, responseText: null, error: `Parse error: ${err instanceof Error ? err.message : String(err)}` })
          }
        })
      })

      request.on('error', (err) => {
        resolve({ imagePath: null, responseText: null, error: err.message })
      })

      request.write(body)
      request.end()
    })
  }
}
